bls/scheme.py

```diff
@@ -132,3 +132,56 @@ def verify(params, aggr_vk, sigma, m):
 	digest = hash(m)
 	h = G.hashG1(digest)
 	return not h.isinf() and e(sigma, g2) == e(h, aggr_vk)
+
+def sign_unique(params, sk, m):
+	"""
+	Sign messages while ensuring the messages will be unique 
+	by appending the public key of the signer with the message.
+
+	Parameters:
+		- `params`: public parameters generated by `setup`
+		- `sk` (Bn): the secret key of the authority
+		- `m` [Bn]: array containing the messages
+
+	Returns:
+		- `sigma_tilde` (G1Elem, G1Elem): blinded credential
+	
+	Notes:
+        To be used with aggregate_verify, where distinct messages are required
+		to prevent rogue public key attacks. 
+	"""
+	assert len(m) > 0
+	(G, o, g1, g2, e) = params
+	vk = sk*g2 # sk*g2 gives the correspongind public key
+	m.append(vk.export()) 
+	digest = hash(m) 
+	h = G.hashG1(digest)
+	sigma = sk*h
+	return sigma
+
+
+def aggregate_verify(params, vks, sigma, m):
```

Comment:
Should we ensure that all `vks` are different? The trick used in `signed_unique` does not seem to work in the case where the same user signs more than one of the messages to aggregate.

Reply:
Right. I didn't notice this particular case. I'll add the code for making sure all `vks` are different.